### DISCUSSION TITLE
Populate hostRules in self-hosting environment

### DISCUSSION THREAD

**Discussion by augi** [renovatebot/renovate#24790] [Request Help]
**What would you like Renovate to be able to do?**

We are running `renovate/renovate:24.113` in k8s, pointing to our GitHub Enterprise installation, and we would like to use the `git-submodule` feature. 

With the initial setup, there is `Error mapping git submodules during extraction` error with `fatal: could not read Username for 'https://our_github_instance': No such device or address` message.

To overcome this issue, we have to set `RENOVATE_HOST_RULES` to `[{"domainName":"our_ghe_installation","token":"same_as_RENOVATE_TOKEN"}]`.

**Did you already have any implementation ideas?**
When debugging this issue, I noticed that `hostRules` is automatically prefilled from `GITHUB_COM_TOKEN`, so it contains following: `{"hostType": "github", "domainName": "github.com", "token": "same_as_GITHUB_COM_TOKEN"}`

So it would IMHO improve usability if `hostRules` were also prefilled from `RENOVATE_ENDPOINT` and `RENOVATE_TOKEN` envvars, when present 🙏

**Comment by rarkins** [comment]
This should actually be done, as part of platform initialization: https://github.com/renovatebot/renovate/blob/34d64832576daf48a04fe34a6f460f15a681c06c/lib/platform/index.ts#L116-L128

You can see that it configures using hostName and not domainName. Would that make a difference to you? BTW domainName should be an actual domain name (e.g. foo.com) and not a full host like foo.bar.com.

**Comment by augi** [comment]
The code looks good but unfortunately, there is something wrong somewhere.

We are executing the container with these envvars specified (the most relevant first):
```
RENOVATE_PLATFORM
RENOVATE_ENDPOINT
RENOVATE_TOKEN
RENOVATE_GIT_AUTHOR
GITHUB_COM_TOKEN
RENOVATE_PRINT_CONFIG
RENOVATE_LOG_LEVEL
RENOVATE_ONBOARDING
RENOVATE_CLONE_SUBMODULES
RENOVATE_REQUIRE_CONFIG
RENOVATE_AUTODISCOVER
RENOVATE_AUTODISCOVER_FILTER
RENOVATE_PAGINATE_ALL
RENOVATE_ENABLED_MANAGERS
RENOVATE_CONFIG_FILE
```
During the run, renovate prints following configs:
```
...
DEBUG: Env config
       "config": {
         "hostRules": [
           {"hostType": "github", "domainName": "github.com", "token": "***********"}
         ],
...
DEBUG: Combined config
       "config": {
         "git-submodules": {"enabled": true},
         "hostRules": [
           {"hostType": "github", "domainName": "github.com", "token": "***********"}
         ],
...
DEBUG: Full resolved config including presets
...
"hostRules": [],
...
```

So as you can see, the `hostRules` is populated from `GITHUB_COM_TOKEN` envvar only.
Also, this causes the following error:
```
 WARN: Error mapping git submodules during extraction (repository=main-org/main-repo)
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             "--symref",
             "https://our_ghe/an-org/a-repo.git",
             "HEAD"
           ],
           "format": "utf-8"
         },
         "message": "fatal: could not read Username for 'https://our_ghe': No such device or address\n",
         "stack": "Error: fatal: could not read Username for 'https://our_ghe': No such device or address\n\n    at Object.action (/usr/src/app/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:38:28)\n    at PluginStore.exec (/usr/src/app/node_modules/simple-git/src/lib/plugins/plugin-store.ts:24:29)\n    at /usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:109:40\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:106:14)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:83:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)"
       }
```

When we specify `RENOVATE_HOST_RULES` as `[{"domainName":"our_ghe","token":"same_as_RENOVATE_TOKEN"}]` then the error disapears.

**Comment by augi** [comment]
So maybe the problem could be that the `git-submodule` implementation doesn't use the `hostRules` populated by the platform? 🤔

**Comment by rarkins** [comment]
Here is the relevant code: https://github.com/renovatebot/renovate/blob/34d64832576daf48a04fe34a6f460f15a681c06c/lib/manager/git-submodules/extract.ts#L115-L117

We could add some debugging there to confirm if a token is found and added.

**Comment by augi** [comment]
That would be great!

**Comment by rarkins** [comment]
Already added in `24.115.1`: https://github.com/renovatebot/renovate/blob/39cd2c616fe7ef8256b5049f5d1bb7d7dcb976e8/lib/manager/git-submodules/extract.ts#L116-L123

**Comment by augi** [comment]
I can confirm that the `Found hostRules token for submodule` message is shown when I have the `RENOVATE_HOST_RULES` specified, but it is not present when I remove the `RENOVATE_HOST_RULES` envvar.

So it looks like this issue could be still valid 🤔

**Comment by augi** [comment]
Maybe the problem is how the _final_ config is composed 🤔
